### PR TITLE
feat(rust): implement basic router, modify message and route types

### PR DESCRIPTION
**Pull request by robinbudd**
- Message: implement std::io::Read & std::io::Write traits
- Message: update Route type to only contain one-way route, i.e.
           separate onward and return routes
- Router: implement basic router functionality and test suite

### CODE REVIEW

File: implementations/rust/router/src/router.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#![allow(unused)]
+pub mod router {
+    use ockam_message::message::*;
+    use std::collections::HashMap;
+    use std::io::{Error, ErrorKind, Read, Write};
+    use std::net::UdpSocket;
+    use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+    use std::ops::Add;
+    use std::sync::{Arc, Mutex};
+    use std::thread;
```
Comment (mrinalwadhwa) [inline implementations/rust/router/src/router.rs:10]: I think we want the router to be `no_std` and have no knowledge of / dependence on transport protocols /  wire encoding / threading / IO etc. 
https://rust-embedded.github.io/book/intro/no-std.html

File: implementations/rust/router/src/router.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    impl HandleMessage for UdpAddressHandler {
+        fn handle_message(&self, mut m: Box<Message>) -> Result<(), std::io::Error> {
+            println!("In UdpAddressHandler!");
+            Ok(())
+        }
+    }
+
+    pub trait HandleMessage {
```
Comment (mrinalwadhwa) [inline implementations/rust/router/src/router.rs:23]: I'm call this `MessageHandler` in elixir, let's use that same name here.

File: implementations/rust/router/src/router.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            Router {
+                registry: HashMap::new(),
+            }
+        }
+
+        pub fn register_handler(
+            &mut self,
+            handler: Arc<Mutex<Box<dyn HandleMessage + Send>>>,
+            address: &Address,
+        ) -> Result<(), String> {
```
Comment (mrinalwadhwa) [inline implementations/rust/router/src/router.rs:42]: This function, I think, should set a handler by `address type` not by `address`.
The signature in the elixir implementation is:

```
set_message_handler_for_address_type(address_type, handler)
```

File: implementations/rust/router/src/router.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        }
+
+        pub fn route(&mut self, mut m: Box<Message>) -> Result<(), String> {
+            // Pop the first address in the list
+            // If there are no addresses, route to the controller
+            // Controller key is always 0
+            let mut key: u64 = 0;
+            let address: Address;
+            if !m.onward_route.addresses.is_empty() {
+                address = m.onward_route.addresses.remove(0);
```
Comment (mrinalwadhwa) [inline implementations/rust/router/src/router.rs:55]: I think the router shouldn't remove the address.
It just hands over the message to a handler based on the `address type` of the first address in the `onward_route` of the message.

**Comment by mrinalwadhwa** [comment]
@robinbudd the commit here https://github.com/ockam-network/ockam/pull/411/commits/807c8524dc2b47498fecdfe7d642d229235c9a57 has many more things changing than what is reflected in the commit message. Could you break this down into multiple commits to have the commit messages reflect the changes

File: implementations/rust/router/src/router.rs (hunk 1 above)
Comment (robinbudd) [reply implementations/rust/router/src/router.rs:10]: The router has no knowledge of transport. The only reason transport-related stuff is in there is for the tests. The actual router code only knows about addresses and dispatch functions.

What do you mean by "no_std"? If I can't use std, it adds weeks of work to implement OS-independent equivalents for hashmap, threads, and all the transport stuff. What do you propose instead? @mrinalwadhwa

File: implementations/rust/router/src/router.rs (hunk 2 above)
Comment (robinbudd) [reply implementations/rust/router/src/router.rs:23]: ok

File: implementations/rust/router/src/router.rs (hunk 3 above)
Comment (robinbudd) [reply implementations/rust/router/src/router.rs:42]: Ok, it was an incomplete refactor after the design changes from a couple days ago.

File: implementations/rust/router/src/router.rs (hunk 4 above)
Comment (robinbudd) [reply implementations/rust/router/src/router.rs:55]: So the handler removes the address? I think the router should be the only thing messing with the routes. All route management done by the router.

**Review by mrinalwadhwa — APPROVED** [review]

**Pull request merged**